Add new filter plugin: 'to_prettytable'

tgadiev opened the pull request:
##### SUMMARY
This PR adds a new filter plugin, `to_prettytable`, which formats a list of dictionaries into an ASCII table using the PrettyTable library. This plugin can be used to generate formatted tables from Ansible variables, making it useful for reports, notifications, and documentation.

The plugin includes features for column ordering, custom header names, and column alignment configuration. It supports handling of empty data and provides clear error messages for invalid inputs.

##### ISSUE TYPE
- New Module/Plugin Pull Request

##### COMPONENT NAME
prettytable

##### ADDITIONAL INFORMATION
This filter plugin formats a list of dictionaries into a readable ASCII table, with the following capabilities:
- Format lists of dictionaries into neatly aligned tables
- Specify column order
- Customize header names
- Configure column alignments (left, center, right)
- Handle empty data sets
- Comprehensive error handling

Example usage:

```yaml
---
- name: Display a list of users as a table
  vars:
    users:
      - name: Alice
        age: 25
        role: admin
      - name: Bob
        age: 30
        role: user
  debug:
    msg: "{{ users | community.general.to_prettytable }}"

- name: Display a table with custom column ordering
  debug:
    msg: "{{ users | community.general.to_prettytable('role', 'name', 'age') }}"

- name: Display a table with custom headers
  debug:
    msg: "{{ users | community.general.to_prettytable(header_names=['User Name', 'User Age', 'User Role']) }}"

- name: Display a table with custom alignments
  debug:
    msg: "{{ users | community.general.to_prettytable(column_alignments={'name': 'center', 'age': 'right', 'role': 'left'}) }}"

- name: Combine multiple options
  debug:
    msg: "{{ users | community.general.to_prettytable(
        column_order=['role', 'name', 'age'],
        header_names=['Position', 'Full Name', 'Years'],
        column_alignments={'name': 'center', 'age': 'right', 'role': 'left'}) }}"
  ```

Output example:
```shell
+-------+-----+-------+
|  name | age |  role |
+-------+-----+-------+
| Alice |  25 | admin |
|  Bob  |  30 |  user |
+-------+-----+-------+
```

The [prettytable](https://pypi.org/project/prettytable) library is required for this module to function.

bcoca commented:
wouldn't this fit better as a filter?

tgadiev commented:
> wouldn't this fit better as a filter?

Filter has a different use case according to my experience. This implementation is useful for debugging and pretty output in console, as `debug` action plugin does. But we can add such filter later if there's such demand.

bcoca commented:
the filter is more useful in many contexts, and you still get 'debug':
```
- debug: msg={{ data|prettytable}}
```
But now you can also use it to write a template/lineinfile/blockinfile/etc

tgadiev commented:
Added the `to_prettytable` filter plugin as per @bcoca suggestion

tgadiev commented:
ready_for_review

tgadiev commented:
bot_status

tgadiev commented:
ready_for_review

russoz commented on plugins/action/prettytable.py:
```diff
@@ -0,0 +1,146 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2020, quidame <quidame@poivron.org>
```
This needs adjusting.

russoz commented on plugins/action/prettytable.py:
```diff
@@ -0,0 +1,146 @@
+
+from typing import Any, Dict, List, Optional
+
+__metaclass__ = type  # pylint: disable=C0103
```
Not sure this is needed in action plugins at the supported versions of ansible-core, but not sure it does not either. @felixfontein ?

felixfontein replied:
It is no longer needed.

russoz replied:
@tgadiev Would you please remove this line? It is no longer needed

russoz commented on plugins/action/prettytable.py:
```diff
@@ -0,0 +1,146 @@
+            if data is None:  # Only check for None, allow empty list
+                raise AnsibleError("The 'data' parameter is required")
+
+            if not isinstance(data, list):
+                raise AnsibleError(
+                    "Expected a list of dictionaries, got a string"
+                    if isinstance(data, (str, bytes))
+                    else f"Expected a list of dictionaries, got {type(data).__name__}"
+                )
+
+            if data and not all(isinstance(item, dict) for item in data):
+                invalid_item = next(item for item in data if not isinstance(item, dict))
+                raise AnsibleError(
+                    "All items in the list must be dictionaries, got a string"
+                    if isinstance(invalid_item, (str, bytes))
+                    else f"All items in the list must be dictionaries, got {type(invalid_item).__name__}"
+                )
```
This is not wrong, but personally I think it is over-engineered. 

I think something like:
```suggestion
            if not isinstance(data, list):
                raise AnsibleError("Parameter 'data' must be a list")

            if data and not all(isinstance(item, dict) for item in data):
                raise AnsibleError("All elements in 'data' must be dictionaries")
```
would get the job done and it would make the code simpler to read and maintain.

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,183 @@
+DOCUMENTATION = '''
+  name: to_prettytable
+  short_description: Format a list of dictionaries as an ASCII table
+  version_added: "8.0.0"
```
```suggestion
  version_added: "10.6.0"
```

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,183 @@
+          choices: [left, center, right, l, c, r]
+'''
+
+EXAMPLES = '''
+- name: Display a list of users as a table
```
Please use:
```suggestion
EXAMPLES = '''
---
- name: Display a list of users as a table
```

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,183 @@
+            'You need to install "prettytable" Python module to use this filter'
+        )
+
+    if not isinstance(data, list):
+        raise AnsibleFilterError(
+            "Expected a list of dictionaries, got a string"
+            if isinstance(data, string_types)
+            else f"Expected a list of dictionaries, got {type(data).__name__}"
+        )
```
As in the action plugin, I think this could be made leaner.

russoz commented on plugins/modules/prettytable.py:
Errr... there is not actual Ansible module in this file, not sure why it has been added to the PR, and I am not sure that a module for pretty printing tables would make sense.

russoz commented on tests/integration/targets/filter_to_prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,130 @@
+      - combined_table | trim is not none
+      - "combined_table | trim is search('Position')"
+      - "combined_table | trim is search('Full Name')"
+      - "combined_table | trim is search('Years')"
```
Give this is a small test table, why not simply check its entire content? Like:
```suggestion
      - "combined_table | trim is search('Years')"
      - combined_table == expected_table
  vars:
    expected_table: |    # perhaps |- , and I just copied the table from the PR description, the contents should be different
      +----------+----------+------+
      | Position | Full Name| Years|
      +----------+----------+------+
      | Developer|   John   |   30 |
      | Designer |   Sara   |   28 |
      +----------+----------+------+
```

russoz commented on tests/integration/targets/filter_to_prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,130 @@
+      - "combined_table | trim is search('Years')"
+
+# Test empty data
+- name: Test empty data list
+  set_fact:
+    empty_table: "{{ [] | community.general.to_prettytable }}"
+
+- name: Verify empty table output
+  assert:
+    that:
+      - empty_table | trim == "++\n++" | trim
```
For such a simple case, one could argue that the fact is not necessary:
```suggestion
- name: Verify empty table output
  assert:
    that:
      - [] | community.general.to_prettytable | trim == "++\n++" | trim
```

russoz commented on tests/integration/targets/prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,205 @@
+# and should not be used as examples of how to write Ansible roles #
+####################################################################
+
+# Copyright (c) 2020, Pavlo Bashynskyi (@levonet) <levonet@gmail.com>
```
This needs adjusting

russoz reviewed (commented):
Hi @tgadiev thanks for your contribution!

This was a first glance, there are some adjustments needed already, but I might be spotting additional ones in the future.

However, I think the most important adjustment that could be made would be to refactor the "prettytable" code out to a new `module_utils`, then you could make both the action and the filter plugins use the exact same code.

That would reduce code duplication in those plugins and code duplication in the tests.

felixfontein commented:
I'm also wondering what the action plugin is needed for. The filter covers all use-cases mentioned so far.

tgadiev commented:
> I'm also wondering what the action plugin is needed for. The filter covers all use-cases mentioned so far.

Removed action plugin as suggested. Moved all functionality to 'to_prettytable' filter plugin.

tgadiev commented:
* Applied all suggested corrections
* Removed action plugin as duplicating thing
* Implemented integration tests using more clear text tables input
* Simplified validation code

@felixfontein @russoz could you, pls, take a look?

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,181 @@
+        raise AnsibleFilterError(type_error(invalid_item, "all items in the list to be dictionaries"))
+
+    # Handle positional argument column order
+    column_order = kwargs.get('column_order', None)
```
I would use `pop`, that way it's easier to validate that no other arguments have been passed.
```suggestion
    column_order = kwargs.pop('column_order', None)
```

tgadiev replied:
Corrected

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,181 @@
+        raise AnsibleFilterError(type_error(invalid_item, "all items in the list to be dictionaries"))
+
+    # Handle positional argument column order
+    column_order = kwargs.get('column_order', None)
```
You need to validate `column_order`.

tgadiev replied:
Corrected

felixfontein replied:
I don't see any checks for that.

tgadiev replied:
I changed 'kwargs.get' to 'kwargs.pop' as you suggested. Maybe I missed the idea. What kind of validation are you talking about?

felixfontein replied:
Using `pop` removes the fields from `kwargs`, so in the end you see whether there are leftovers that you can report to the user as an error ("specified unknown parameters").

felixfontein replied:
Ah, and validation: you need to make sure that what the user passed as `column_order` is actually what you expect it to be. For example if the user passes `column_order=123`, you have to handle it.

tgadiev replied:
Ok, got it now. Added additional validation and corresponding tests.

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,181 @@
+    field_names = column_order or list(data[0].keys())
+
+    # Set headers
+    header_names = kwargs.get('header_names', None)
```
Same here: use `pop` and validate `header_names`.

tgadiev replied:
Corrected

felixfontein replied:
I don't see checks for this either.

tgadiev replied:
Added additional validation and corresponding tests.

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,181 @@
+    table.field_names = header_names if header_names else field_names
+
+    # Configure alignments
+    _configure_alignments(table, field_names, kwargs.get('column_alignments', {}))
```
Same here.

tgadiev replied:
Corrected

tgadiev replied:
Added additional validation and corresponding tests.

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,181 @@
+
+    # Configure alignments
+    _configure_alignments(table, field_names, kwargs.get('column_alignments', {}))
+
```
At this point you should fali if `kwargs` still contains something.

tgadiev replied:
Corrected

felixfontein replied:
Are you sure? I don't see code that reports an error if `kwargs` still contains something at this point (or later).

tgadiev replied:
Added additional validation and corresponding tests.

felixfontein reviewed (commented):
Unfortunately you have to do all parameter validation yourself for filter plugins...

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,181 @@
+        alignment:
+          description: Alignment for the column. Must be one of C(left), C(center), C(right), C(l), C(c), or C(r).
+          type: string
+          choices: [left, center, right, l, c, r]
```
Basically you're saying: `column_alignments` should be a dictionary containing a single key `alignment` that has one of the values `left`, `center`, ...

Since that's not what you want to say, you shouldn't use `suboptions`. Extend `column_alignment`'s `description` (use multiple paragraphs to make it more readable) to describe the correct input.

tgadiev replied:
Description adjusted. Thanks for notice!

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,181 @@
+
+    # Helper function for type error messages
+    def type_error(obj, expected):
+        type_name = "string" if isinstance(obj, string_types) else type(obj).__name__
```
Why not use the name of the actual type of the object? What do we gain by translating that to `string`?

tgadiev replied:
In Python, especially with Python 2/3 compatibility concerns, string types can be represented by different classes (str, unicode, bytes, etc.). Using "string" simplifies this complexity for the error message recipient.

felixfontein replied:
Since we don't support Python 2 anymore for plugins other than modules this won't be a problem, but Ansible has several internal types for strings (and with data tagging, also for other types) that can be confusing for users. So this is OK.

russoz replied:
> In Python, especially with Python 2/3 compatibility concerns, string types can be represented by different classes (str, unicode, bytes, etc.). Using "string" simplifies this complexity for the error message recipient. 

I don't think it is a good thing, for alnost the exact reasons you think it is. This message is going to be used by someone debugging their code, and in that context, having more information is usually better than having less.

Another argument against that is consistency: I have not searched nor really intend to at thus point, but I believe that other plugins don't do that "translation" - at least I don't recall seeing that before. But I might be wrong. 

That being said, I'm not going to press the point, as I seem to be the minority here.

russoz commented on tests/integration/targets/filter_to_prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,166 @@
+- name: Verify basic table output
+  assert:
+    that:
+      - basic_table | trim == expected_basic_table | trim
```
Why are we testing the output piped into `trim`? That could potentially hide some inconsistency in the output of the filter.

tgadiev replied:
Agree. Removed `trim` filter from tests.

tgadiev commented:
ready_for_review

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,180 @@
+      type: list
+      elements: string
+    column_alignments:
+      description: >-
+        Dictionary where keys are column names and values are alignment settings.
+        Valid alignment values are C(left), C(center), C(right), C(l), C(c), or C(r).
+        For example, {'name': 'left', 'id': 'right'} will align the 'name' column to the left
+        and the 'id' column to the right.
```
1. I'd split this up into two paragraphs. That makes it much more readable.
2. Use formatting for the example.
```suggestion
      description:
        - >-
          Dictionary where keys are column names and values are alignment settings.
          Valid alignment values are C(left), C(center), C(right), C(l), C(c), or C(r).
        - >-
          For example, V({'name': 'left', 'id': 'right'}) will align the C(name) column to the left
          and the C(id) column to the right.
```

tgadiev replied:
Corrected as requested.

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,190 @@
+    # Handle positional argument column order
+    column_order = kwargs.pop('column_order', None)
+    if column_order is not None and not isinstance(column_order, list):
+        raise AnsibleFilterError(type_error(column_order, "a list of column names"))
```
Shouldn't you also check that the elements in `column_order` are strings?

tgadiev replied:
Added necessary checks and integration tests.

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,190 @@
+    if column_order is not None and not isinstance(column_order, list):
+        raise AnsibleFilterError(type_error(column_order, "a list of column names"))
+    if args and not column_order:
+        column_order = list(args)
```
You also have to check the entries of `args`.

Also I'd fail if both `args` and `column_order` are provided, instead of silently ignoring `args` in that case.

tgadiev replied:
Added necessary checks and integration tests.

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,190 @@
+    # Set headers
+    header_names = kwargs.pop('header_names', None)
+    if header_names is not None and not isinstance(header_names, list):
+        raise AnsibleFilterError(type_error(header_names, "a list of header names"))
```
You need to check the entries of `header_names`. Also does this list should have the same length as `field_names`?

tgadiev replied:
I prefer to make it more flexible. Both the `column_order` and `header_names` parameters can include selected keys and their corresponding custom headers. Therefore, the size of `column_order` and `header_names` can be smaller than the size of the input data dictionaries, but not larger. Additionally, if both parameters are provided, their sizes must match. I’ve added the necessary checks and tests for these cases

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,190 @@
+    table = prettytable.PrettyTable()
+
+    # Determine field names
+    field_names = column_order or list(data[0].keys())
```
The dictionary keys are not necessarily strings. Since likely strings are needed, you need to convert them.

tgadiev replied:
Added necessary conversion and tests for numeric and boolean keys if they occur in input data.

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,190 @@
+
+    # Check for unknown parameters
+    if kwargs:
+        raise AnsibleFilterError(f"Unknown parameter(s) for to_prettytable filter: {', '.join(kwargs.keys())}")
```
I'd sort the keys. Also `.keys()` is not necessary, since iterating over a dictionary iterates over its keys.
```suggestion
        raise AnsibleFilterError(f"Unknown parameter(s) for to_prettytable filter: {', '.join(sorted(kwargs))}")
```

tgadiev replied:
Done

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,190 @@
+
+    for col_name, alignment in column_alignments.items():
+        if col_name in field_names:
+            alignment = alignment.lower()
```
Why convert the values to lower-case?

Also you first have to check whether they are actually strings, and if they're strings, that they're one of `valid_alignments`.

tgadiev replied:
Added necessary validation and corresponding tests.

russoz commented on tests/integration/targets/filter_to_prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,202 @@
+
+- name: Remove copyright header from expected output
+  set_fact:
+    expected_basic_table: "{{ expected_basic_table_raw.split('\n')[4:] | join('\n') }}"
```
As mentioned in a previous comment, I think it would be easier to add the resulting tables literally in here, something like:
```suggestion
    expected_basic_table: |
        +-------+-----+-------+
        |  name | age |  role |
        +-------+-----+-------+
        | Alice |  25 | admin |
        |  Bob  |  30 |  user |
        +-------+-----+-------+
```
It saves you the trouble of reading files and cleaning the licensing lines.

That being said, if you _really want_ (I see little to gain by doing that) to keep the external files, I would strongly recommend you remove the license header from them, create an adjacent file - in this example `basic_table.txt.license`, add the licensing information there so that the data file is kept clean and you don't need to remove this arbitrary number of lines from the content.

Naturally, the same argument applies to the other files as well.

tgadiev replied:
Done

russoz commented on tests/integration/targets/filter_to_prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,466 @@
+      assert:
+        that:
+          - failure_result is failed
+          - "\"'header_names' has more elements\" in failure_result.msg"
```
This is probably more readable as
```suggestion
          - >
            "'header_names' has more elements" in failure_result.msg
```

tgadiev replied:
Done. Nice suggestion!

russoz commented on tests/integration/targets/filter_to_prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,466 @@
+  set_fact:
+    empty_table: "{{ [] | community.general.to_prettytable }}"
+    expected_empty_table: |-
+      ++
+      ++
```
ha! so that's what an empty table looks like :-P

tgadiev commented:
ready_for_review

felixfontein commented:
No idea, it's broken in all repos that use it to my knowledge. In any case, restarting is only wasting CPU resources, so please don't do it unnecessarily.

tgadiev commented:
ready_for_review

tgadiev commented:
I added all suggested changes and made a rebase to update CI tests to actual state.

@felixfontein @russoz, can you please review?

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,382 @@
+
+    # === Process column order ===
+    # Handle both positional and keyword column_order
+    column_order = kwargs.pop('column_order', None)
+
+    # Check for conflict between args and column_order
+    if args and column_order is not None:
+        raise AnsibleFilterError("Cannot use both positional arguments and the 'column_order' keyword argument")
+
+    # Use positional args if provided
+    if args:
+        column_order = list(args)
```
This could be simpler as:
```suggestion
    if args and column_order is not None:
        raise AnsibleFilterError("Cannot use both positional arguments and the 'column_order' keyword argument")

    column_order = args or kwargs.pop('column_order', None)
```
(btw, `args` is already a list, no need to put it through `list()`)

tgadiev replied:
Sorry, it looks like the suggested approach does not work as expected. First, you cannot check `column_order` value in line 258 before you really set this variable in line 261. It returns an error of undefined variable this way. Second, removing explicit list in line 266 (from initial code) breaks integration tests and makes it impossible to check the args correctly. If I remove explicit list from line 266, I get the following error in tests:
```
TASK [filter_to_prettytable : set_fact] ****************************************
fatal: [testhost]: FAILED! => {"msg": "Expected a list of column names, got a tuple"}
...ignoring

TASK [filter_to_prettytable : Verify error message for non-string values in positional args] ***
fatal: [testhost]: FAILED! => {
    "assertion": "'Expected a string for column name, got a int' in failure_result.msg",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,382 @@
+        raise AnsibleFilterError(_type_error(invalid_item, "all items in the list to be dictionaries"))
+
+    # Get the maximum number of fields in the first dictionary
+    max_fields = len(data[0].keys())
```
The `.keys()` is redundant here.
```suggestion
    max_fields = len(data[0])
```

tgadiev replied:
Done

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,382 @@
+        field_names = column_order
+    else:
+        # Use field names from first dictionary, ensuring all are strings
+        field_names = [to_text(k) for k in data[0].keys()]
```
redundant
```suggestion
        field_names = [to_text(k) for k in data[0]]
```

tgadiev replied:
Done

russoz commented on tests/integration/targets/filter_to_prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,520 @@
+# Test column ordering with numeric keys
+- name: Test column ordering with numeric keys
+  set_fact:
+    mixed_ordered_table: "{{ mixed_key_data | community.general.to_prettytable(column_order=['1', 'name', 'role']) }}"
```
Is it going to work if used as:
```suggestion
    mixed_ordered_table: "{{ mixed_key_data | community.general.to_prettytable(column_order=[1, 'name', 'role']) }}"
```
?
Should it? Should we have a separate test for that?

tgadiev replied:
No, it shouldn’t work this way. We have a strict requirement that all `column_order` elements must be strings. Validation for this is already implemented — see the `_validate_list_param` function in the filter code. This case is also covered by our integration tests. You can check that part there
```yaml
- name: Test non-string values in column_order (expect failure)
  block:
    - set_fact:
        invalid_table: "{{ test_data | community.general.to_prettytable(column_order=['name', 123, 'role']) }}"
      register: failure_result
      ignore_errors: true
    - name: Verify error message for non-string values in column_order
      assert:
        that:
          - failure_result is failed
          - "'Expected a string for column name, got a int' in failure_result.msg"
```

russoz reviewed (commented):
hi @tgadiev 

I spotted a couple of things that should/could/might be adjusted. Take it with a pinch of salt.

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,382 @@
+from __future__ import (absolute_import, division, print_function)
+
+DOCUMENTATION = '''
+  name: to_prettytable
```
Also, for consistency with other plugins, please do not indent the top level of the documentation blocks.
```suggestion
name: to_prettytable
```
(replicate throughout this and the other doc blocks, please)

tgadiev replied:
Done

tgadiev commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,382 @@
+        Formatted error message string
+    """
+    type_name = "string" if isinstance(obj, string_types) else type(obj).__name__
+    return f"Expected {expected}, got a {type_name}"
```
Done

tgadiev commented:
ready_for_review

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,380 @@
+'''
+
+RETURN = '''
+  _value:
+    description: The formatted ASCII table.
+    type: string
```
```suggestion
_value:
  description: The formatted ASCII table.
  type: string
```

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,380 @@
+            'You need to install "prettytable" Python module to use this filter'
+        )
+
+    # Handle empty data
+    if not data:
+        return "++\n++"
```
I'd move this below the validation of the options.

tgadiev replied:
I tested such approach. If we process empty data after other options and data processing, it causes exception: `IndexError: list index out of range`

tgadiev replied:
Some parameter validations depend on the input data type. We validate the size of certain parameters, such as `column_order` and `header_names`, by comparing them to the number of fields in the input dictionary. If we process empty data after that, it may lead to an exception like the one described above.

felixfontein replied:
Now this can be moved down after the argument validation, I think.

tgadiev replied:
If we change the code this way, it will make it impossible to use the filter with custom header parameters while still allowing dynamic processing of input data (including zero-length lists). As a result, if we receive an empty list as input (which can happen in some cases) and `header_names` and `column_order` parameters are set, we would always get an error instead of a zero-length table. I believe this would make the filter less convenient to use.

felixfontein replied:
Then please adjust these specific validations accordingly. Simply not validating any other parameter **at all** when `data` is empty is wrong.

tgadiev replied:
Adjusted the code to handle validations properly even with empty input data. Added integration tests for new use cases with empty data.

russoz commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,380 @@
+- name: Display a list of users as a table
+  vars:
+    users:
+      - name: Alice
+        age: 25
+        role: admin
+      - name: Bob
+        age: 30
+        role: user
```
The consistency pedantic strikes again 
```suggestion
    - name: Alice
      age: 25
      role: admin
    - name: Bob
      age: 30
      role: user
```

tgadiev replied:
Sorry, what consistency are we talking about here? This is an example of an ansible code in yaml format. And according to yamllint rules (used by ansible-lint for example) block sequences should be indented by default. See here for example:
https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.indentation

I'm not against this change, but I'm really confused and I don't understand why we need it. I checked other filter plugins code and I can see the same indentation as mine there. For example, `lists_mergeby`.

felixfontein replied:
The indentation in the PR looks good to me, I'm not sure why it should be changed. The resulting indentation doens't look correct.

russoz replied:
First of all, I'm not holding this back because of indentation, I don't want to make this thing bigger than it is. That being said, I apologise for not being very clear. The consistency difference I was referring to is the one between a list like lines 15-16
```yaml
requirements:
- prettytable
```
and lines 46-47
```yaml
    users:
      - name: Alice
```
The `-` is aligned with the containing element in one, whereas in the other it is indented 2 spaces.

Again, this is a very tiny nit, this is not holding back the merge, I just wanted to explain I wrote that before.

felixfontein replied:
In that case lines 15-16 should be fixed, IMO, as generally the indentation in this repo is matching the quoted block, but not lines 15-16.

tgadiev replied:
Fixed

russoz replied:
Thanks!

russoz reviewed (commented):
One more

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,380 @@
+    elements: string
+  column_alignments:
+    description:
+    - >-
+        Dictionary where keys are column names and values are alignment settings.
+        Valid alignment values are C(left), C(center), C(right), C(l), C(c), or C(r).
+    - >-
+        For example, V({'name': 'left', 'id': 'right'}) will align the C(name) column to the left
+        and the C(id) column to the right.
```
This one is off though:
```suggestion
      - >-
          Dictionary where keys are column names and values are alignment settings.
          Valid alignment values are C(left), C(center), C(right), C(l), C(c), or C(r).
      - >-
          For example, V({'name': 'left', 'id': 'right'}) will align the C(name) column to the left
          and the C(id) column to the right.
```

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,380 @@
+
+- name: Combine multiple options
+  debug:
+    msg: "{{ users | community.general.to_prettytable(
+        column_order=['role', 'name', 'age'],
+        header_names=['Position', 'Full Name', 'Years'],
+        column_alignments={'name': 'center', 'age': 'right', 'role': 'left'}) }}"
```
Instead of using a multi-line string with quotes, it's better to use `>-`. I would indent it this way, to increase readability:
```suggestion
    msg: >-
      {{
        users | community.general.to_prettytable(
          column_order=['role', 'name', 'age'],
          header_names=['Position', 'Full Name', 'Years'],
          column_alignments={'name': 'center', 'age': 'right', 'role': 'left'},
        )
      }}
```

tgadiev replied:
Revised the example tasks' style to match the proposed variant.

felixfontein commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,380 @@
+---
+- name: Display a list of users as a table
+  vars:
+    users:
```
Since this variable is used in all examples, either use a `block:`, or use `ansible.builtin.set_fact` to store it as a fact before showing the examples.

tgadiev replied:
Done

tgadiev commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,409 @@
+        field_names = column_order
+    else:
+        # Use field names from first dictionary, ensuring all are strings
+        field_names = [to_text(k) for k in data[0]]
```
Sure, but why would we need extra code to achieve the same result?

tgadiev commented on plugins/filter/to_prettytable.py:
```diff
@@ -0,0 +1,409 @@
+        raise TypeValidationError(invalid_item, "all items in the list to be dictionaries")
+
+    # Get the maximum number of fields in the first dictionary
+    max_fields = len(data[0])
```
From my perspective, fewer lines of code are preferable when functionality remains unchanged.

felixfontein replied:
Right now argument valiation is defective since it only happens if `data` is not empty. This fixes that defect.

tgadiev replied:
I don’t consider this a defect because the empty data case is already handled earlier in the code (line 267):
```pyhon
    # Handle empty data
    if not data:
        return "++\n++"
```

felixfontein replied:
It is a defect, because invalid input for other options is not caught when `data` happens to be empty.

russoz commented on tests/integration/targets/filter_to_prettytable/tasks/main.yml:
```diff
@@ -0,0 +1,658 @@
+# Test empty data
+- name: Test empty data list with no parameters
+  set_fact:
+    empty_table: "{{ [] | community.general.to_prettytable }}"
```
To test those conditions that @felixfontein was talking about, it _might_ be interesting to add one (or more, if you think it might help) test case with empty data but extra parameters

tgadiev replied:
I've already done that.

russoz reviewed (approved):
LGTM

felixfontein commented:
@tgadiev thanks for your contribution!
@bcoca @russoz thanks for reviewing!
